### PR TITLE
Grid View with variable column widths

### PR DESCRIPTION
**Pull request by vitvakatu**
### Pull Request Description

[ci no changelog needed]
[Task link](https://www.pivotaltracker.com/story/show/182955595)

This PR implements variable column widths in the new Grid View component. We need this feature to quickly implement various parts of the UI, including the breadcrumbs panel of the component browser.

There are two ways to change the width of the specific column:
1. "From the outside", using the `set_column_width` endpoint of the Grid View
2. "From the inside", using the `override_column_width` endpoint of the EntryFrp.

Both ways work similarly, but the latter is helpful for our breadcrumbs implementation, as it allows for entry to decide on the width of the column by its content.

See the screencast with three grid views. The top-left one has every even column shrunk by GridView API. Every grid view has a second column extended by EntryFrp API.

https://user-images.githubusercontent.com/6566674/185060985-7b7df076-c659-41fa-977a-22875493f8d4.mp4

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 1]
```diff
@@ -282,15 +391,37 @@ impl<Entry: display::Object, EntryParams> Model<Entry, EntryParams> {
             entry
         });
         free_entries.extend(detached);
-        let uncovered = all_visible_locations(&viewport, entries_size, row_count, col_count)
+        let widths = &self.column_widths;
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]: I'd put this alias at the function beginning, and use it in line 385.

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 2]
```diff
@@ -263,17 +371,18 @@ impl<Entry, EntryParams> Model<Entry, EntryParams> {
...
         let visible_entries = default();
         let free_entries = default();
-        Model { display_object, visible_entries, free_entries, entry_creation_ctx }
+        let column_widths = ColumnWidths::new(0);
+        Model { display_object, visible_entries, free_entries, entry_creation_ctx, column_widths }
     }
 }
 
-impl<Entry: display::Object, EntryParams> Model<Entry, EntryParams> {
+impl<Entry: display::Object + entry::Entry, EntryParams> Model<Entry, EntryParams> {
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]: The `entry::Entry` implies `display::Object`. Also, if we require `entry::Entry`, the `EntryParams` argument should be `Entry::Params`, not a generic one. And, there is already a block with entry::Entry bounds.

But in the first place, make sure that every function in this block require Entry bound. I have impression, that update_entries_visibility does not. We should try to require trait bounds only in functions where it's necessary.

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 3]
```diff
@@ -303,14 +434,28 @@ impl<Entry: display::Object, EntryParams> Model<Entry, EntryParams> {
...
+        properties.all_visible_locations(&self.column_widths).collect_vec()
     }
 
     fn drop_all_entries(&self, properties: Properties) -> Vec<(Row, Col)> {
         let to_model_request = self.reset_entries(properties);
         self.free_entries.borrow_mut().clear();
         to_model_request
     }
+
+    fn resize_column(&self, col: Col, width: f32, properties: Properties) {
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs:348]: In this function we take only entry_size from properties, so we could take it directly I think.

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 4]
```diff
@@ -323,23 +468,28 @@ impl<E: Entry> Model<E, E::Params> {
...
+                Occupied(entry) => entry.into_mut(),
+                Vacant(lack_of_entry) => {
+                    let new_entry = free_entries.pop().unwrap_or_else(create_new_entry);
+                    set_entry_position(&new_entry, row, col, entry_size, &self.column_widths);
+                    new_entry.entry.frp().set_location((row, col));
+                    self.display_object.add_child(&new_entry);
+                    lack_of_entry.insert(new_entry)
+                }
+            };
+            entry.entry.frp().clone_ref()
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]: Good catch! There is still one FRP event emitted during borrow, in line 483. I wonder how to do that - maybe we can make this block returning entry_frp and a flag indicating if the entry was reassigned.

Also, add a comment why the block is needed.

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 5]
```diff
@@ -147,6 +150,87 @@ impl<E: display::Object> display::Object for VisibleEntry<E> {
...
+        let old_length = borrowed.len();
+        let length = col_count;
+        let double_length = length * 2;
+        if old_length < length {
+            let mut new_vec = Vec::with_capacity(double_length);
+            new_vec.resize(double_length, 0.0);
+            new_vec[length..length + old_length].copy_from_slice(borrowed.view());
+            *borrowed = segment_tree_noalloc(new_vec);
+        }
+    }
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]: Have you seen [`PrefixPoint`](https://docs.rs/segment-tree/2.0.0/segment_tree/struct.PrefixPoint.html) from the same crate? We meet the conditions, and it allows for quick resizing.

File: lib/rust/ensogl/component/grid-view/src/visible_area.rs [hunk 6]
```diff
@@ -31,12 +32,24 @@ pub fn visible_rows(v: &Viewport, entry_size: Vector2, row_count: usize) -> Rang
...
+    entry_size: Vector2,
+    col_count: usize,
+    column_widths: &ColumnWidths,
+) -> Range<Col> {
+    let column_left = |idx: usize| idx as f32 * entry_size.x + column_widths.pos_offset(idx);
+    let column_right = |idx: usize| column_left(idx + 1);
+    let index = |(idx, _): (usize, _)| idx;
+
+    let visible = |(_, x): &(_, f32)| v.left < *x;
+    let mut right_borders = (0..col_count).map(|idx| (idx, column_right(idx)));
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/visible_area.rs]: I don't like iterating over all columns each time the viewport changes - it may change every frame. At least in the case where no column resizes occurs, it should be as fast as before. So let's count the "first_not_visible" as before and check the position offset; if it's big, then jump accordingly.

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 7]
```diff
@@ -147,6 +150,87 @@ impl<E: display::Object> display::Object for VisibleEntry<E> {
 
 
 
+// ====================
+// === ColumnWidths ===
+// ====================
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]: When implementing this crate, I try an approach to keep very focused modules. So, please extract this section to `column_widths` module.

**Review by farmaazon — CHANGES_REQUESTED** [review]
The blocker is a new way of counting visible columns.

File: lib/rust/ensogl/component/grid-view/src/lib.rs (hunk 5 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/lib.rs]: Do you mean the combination of [`unwrap`](https://docs.rs/segment-tree/2.0.0/segment_tree/struct.PrefixPoint.html#method.unwrap) + [`build`](https://docs.rs/segment-tree/2.0.0/segment_tree/struct.PrefixPoint.html#method.build) as "quick resizing"? That would work.

However `PrefixPoint` has O(log i) complexity for [`get`](https://docs.rs/segment-tree/2.0.0/segment_tree/struct.PrefixPoint.html#method.get), comparing to the similar method of `SegmentPoint`. And we need this getter for proper positioning, unfortunately. It can be replaced by `query(idx + 1) - query(idx)` though.

Overall, I think replacing the storage for `PrefixPoint` makes sense. Thanks.

File: lib/rust/ensogl/component/grid-view/src/visible_area.rs (hunk 6 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/visible_area.rs]: I'm a bit confused. You refer to the `first_visible`-related code but mention `first_not_visible` in the comment. What do you mean by "as before"? `let first_visible = v.left / entry_size.x` is not possible anymore, because `entry_size.x` is variable now. There are cases when the first visible column is placed after some extended column, and we need to look through every previous column. 
I guess we can calculate the column's index that *should* be there and then check its `pos_offset`, and if it has any, then retreat back to the iterative solution. Is that your idea?

Comment (farmaazon) [reply lib/rust/ensogl/component/grid-view/src/visible_area.rs]: > I'm a bit confused. You refer to the `first_visible`-related code but mention `first_not_visible` in the comment. What do you mean by "as before"? `let first_visible = v.left / entry_size.x` is not possible anymore, because `entry_size.x` is variable now. There are cases when the first visible column is placed after some extended column, and we need to look through every previous column. I guess we can calculate the column's index that _should_ be there and then check its `pos_offset`, and if it has any, then retreat back to the iterative solution. Is that your idea?

Yes, more or less. What I imagined is start for searching "first visible column" from the old "first_visible" value. And we check the current actual column position: if the column is still on the left viewport boundary, we're good. If not, we jump c.a. `(position_offset / entry_size.x) as isize` columns and check the new one.

This solution is not worse than iterative, but in case where only small amount of columns is resized it should look up for proper column very quickly

There is also better solution, but it requires changes in the SegmentPoint implementation: knowing the x you can traverse the tree from the root, going to the left children if the sum in it is less than x, and right child otherwise.

File: lib/rust/ensogl/component/grid-view/src/lib.rs (hunk 1 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/lib.rs]: Done

File: lib/rust/ensogl/component/grid-view/src/lib.rs (hunk 2 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/lib.rs]: I moved it to the next impl block, you're right.

File: lib/rust/ensogl/component/grid-view/src/lib.rs (hunk 3 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/lib.rs:348]: We use properties on the very next line (first line of the function).

File: lib/rust/ensogl/component/grid-view/src/lib.rs (hunk 4 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/lib.rs]: Fixed & added a comment

File: lib/rust/ensogl/component/grid-view/src/lib.rs (hunk 5 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/lib.rs]: Replaced with `PrefixPoint`

File: lib/rust/ensogl/component/grid-view/src/visible_area.rs (hunk 6 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/visible_area.rs]: I tried the proposed solution but didn't manage to finish it. There are a lot of subtle corner cases, and the resulting implementation is complicated. (and didn't pass the unit tests)
I noticed that with small position offsets, the solution is not better than iterative because we usually jump by only one column at a time.

So I retreated to the dumb guess + iterative solution if the found column has a non-zero position offset.

File: lib/rust/ensogl/component/grid-view/src/lib.rs (hunk 7 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/lib.rs]: Done

File: lib/rust/ensogl/component/grid-view/src/column_widths.rs [hunk 8]
```diff
@@ -0,0 +1,91 @@
+//! Storage for column's widths in the GridView. See [`ColumnWidths`].
+
+use crate::prelude::*;
+use std::fmt::Formatter;
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/column_widths.rs]:
```suggestion
use crate::prelude::*;

use std::fmt::Formatter;
```
I think it could be fixed with `cargo run -p enso-formatter`.

File: lib/rust/ensogl/component/grid-view/src/column_widths.rs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+
+
+// ====================
+// === ColumnWidths ===
+// ====================
+
+
+// === Helper definitions ===
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/column_widths.rs]:
```suggestion
// ====================

// === Helper definitions ===
```
We don't add additional space between headers.

File: lib/rust/ensogl/component/grid-view/src/column_widths.rs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    segment_tree::PrefixPoint::build(width_diffs, segment_tree::ops::Add)
+}
+
+
+// === ColumnWidths ===
+
+/// Storage for column widths.
+///
+/// It stores not the actual widths but rather the differences between [`Frp::set_entry_size`] and
+/// the width updated through [`Frp::set_column_width`] or [`EntryFrp::override_column_width`].
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/column_widths.rs]: `set_entry_size` is a size, so it should rather go "between the default width set by [`Frp::set_entry_size`] endpoint"

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 11]
```diff
@@ -472,24 +374,48 @@ impl<E: Entry> Model<E, E::Params> {
...
+        &self,
+        resized_column: Col,
+        properties: Properties,
+    ) -> Vec<(Row, Col)> {
+        let to_model_request = self.update_entries_visibility(properties);
+        for ((row, col), visible_entry) in &*self.visible_entries.borrow() {
+            // We are not interested in columns to the left of the resized column.
+            if *col < resized_column {
+                continue;
+            }
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]: We can also use `filter` method of Iterator. This way we avoid jumps inside the blocks.

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 12]
```diff
@@ -472,24 +374,48 @@ impl<E: Entry> Model<E, E::Params> {
...
+            let size = properties.entries_size;
+            set_entry_position(visible_entry, *row, *col, size, &self.column_widths);
+            let width_diff = self.column_widths.width_diff(*col);
+            visible_entry.entry.frp().set_size(size + Vector2(width_diff, 0.0));
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]: Emitting events during borrow.

I think we can't avoid making additional `Vec` here: we should just map `self.visible_entries.borrow().iter()` to pairs (visible_entry, size), collect to vector and then iterate over it.

File: lib/rust/ensogl/component/grid-view/src/visible_area.rs [hunk 13]
```diff
@@ -38,18 +38,39 @@ pub fn visible_columns(
...
     let index = |(idx, _): (usize, _)| idx;
-
-    let visible = |(_, x): &(_, f32)| v.left < *x;
-    let mut right_borders = (0..col_count).map(|idx| (idx, column_right(idx)));
-    let first_visible = right_borders.find(visible).map(index).unwrap_or(col_count);
-
+    let clamp = |idx: isize| idx.clamp(0, col_count as isize) as Col;
+
+    let left_guess = clamp((v.left / entry_width).floor() as isize);
+    let first_visible = {
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/visible_area.rs]: Add an explanation why we do such guesses.

File: lib/rust/ensogl/component/grid-view/src/visible_area.rs [hunk 14]
```diff
@@ -38,18 +38,39 @@ pub fn visible_columns(
...
+        let pos_offset = column_widths.pos_offset(left_guess);
+        let column_left_border = left_guess as f32 * entry_width + pos_offset;
+        let column_width = entry_width + column_widths.width_diff(left_guess);
+        let column_right_border = column_left_border + column_width;
+        let is_visible = column_right_border > v.left && column_left_border < v.right;
+        let not_shifted = pos_offset == 0.0;
+
+        if not_shifted && is_visible {
+            left_guess
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/visible_area.rs]: Actually, the left_guess could be shifted and still on the edge. Instead of `is_visible` checking, I would check if `column_left_border <= v.left && column_right_border >= v.left`.

Or, if you prefer simple code, we don't have to check visibility? I deem If it's not shifted, it's always visible.

File: lib/rust/ensogl/component/grid-view/src/visible_area.rs [hunk 15]
```diff
@@ -38,18 +38,39 @@ pub fn visible_columns(
     col_count: usize,
     column_widths: &ColumnWidths,
 ) -> Range<Col> {
-    let column_left = |idx: usize| idx as f32 * entry_size.x + column_widths.pos_offset(idx);
+    let entry_width = entry_size.x;
+    let column_left = |idx: usize| idx as f32 * entry_width + column_widths.pos_offset(idx);
     let column_right = |idx: usize| column_left(idx + 1);
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/visible_area.rs]: Technically, we can call `column_widths.pos_offset(idx);` with the `idx` being equal to col_cound. Does our structure allow that? If yes, add the information about the fact to the `pos_offset` method docs.

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: lib/rust/ensogl/component/grid-view/src/visible_area.rs (hunk 14 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/visible_area.rs]: It can be shifted to the right and be still visible, but it won't be the **first** visible column. A proposed `column_left_border <= v.left && column_right_border >= v.left` doesn't work for the same reason.

> I deem If it's not shifted, it's always visible.

This is also not correct, unfortunately. If a column is shrunk, its position is unchanged (it has pos_offset of zero), but it still can be behind the left viewport border in some cases.

File: lib/rust/ensogl/component/grid-view/src/column_widths.rs (hunk 8 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/column_widths.rs]: Fixed, thanks

File: lib/rust/ensogl/component/grid-view/src/column_widths.rs (hunk 9 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/column_widths.rs]: Fixed, though I wish `enso-formatter` would handle that. (it does not at the moment)

File: lib/rust/ensogl/component/grid-view/src/lib.rs (hunk 12 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/lib.rs]: Done as proposed

File: lib/rust/ensogl/component/grid-view/src/visible_area.rs (hunk 13 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/visible_area.rs]: Added

File: lib/rust/ensogl/component/grid-view/src/visible_area.rs (hunk 15 above)
Comment (vitvakatu) [reply lib/rust/ensogl/component/grid-view/src/visible_area.rs]: Yes, we can pass `idx == col_count`. Extended documentation of that method.

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 16]
```diff
@@ -391,12 +391,15 @@ impl<E: Entry> Model<E, E::Params> {
             };
             (entry.entry.frp().clone_ref(), should_set_location)
         };
-        let width_offset = self.column_widths.width_diff(col);
-        entry_frp.set_size(entry_size + Vector2(width_offset, 0.0));
-        entry_frp.set_model(model);
+        // We should update the location as soon as possible because it is used when handling other
+        // FRP events. E.g. `set_model` can trigger `override_column_width` which needs up to date
+        // location.
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]: My proposition of better comment is:
> The location should be updated first, because computing entry position after column width change uses information about it. And column width may be change as a reaction of any other event, depending of `Entry` implementation.

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 17]
```diff
@@ -405,15 +408,23 @@ impl<E: Entry> Model<E, E::Params> {
...
-            let width_diff = self.column_widths.width_diff(*col);
-            visible_entry.entry.frp().set_size(size + Vector2(width_diff, 0.0));
+            let entries = borrowed.iter().filter(|((_, col), _)| *col >= resized_column);
+            let entries_and_sizes = entries.map(|((row, col), entry)| {
+                let size = properties.entries_size;
+                set_entry_position(entry, *row, *col, size, &self.column_widths);
+                let width_diff = self.column_widths.width_diff(*col);
+                (entry.clone_ref(), size + Vector2(width_diff, 0.0))
+            });
+            entries_and_sizes.collect::<Vec<_>>()
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]:
```suggestion
            entries_and_sizes.collect_vec()
```

File: lib/rust/ensogl/component/grid-view/src/visible_area.rs [hunk 18]
```diff
@@ -44,6 +44,10 @@ pub fn visible_columns(
     let index = |(idx, _): (usize, _)| idx;
     let clamp = |idx: isize| idx.clamp(0, col_count as isize) as Col;
 
+    // We guess the first visible column by the position of the viewport. If there are no resized
+    // columns, that would be the answer. If the guessed column is shifted by resized columns (or is
+    // resized itself), then we iterate over all columns to find the correct one. We repeat a
+    // similar process for the last visible column.
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/visible_area.rs]:
```suggestion
    // We guess the first visible column by the position of the viewport. If there are no resized
    // columns (a common scenario), that would be the answer. If the guessed column is shifted by resized columns (or is
    // resized itself), then we iterate over all columns to find the correct one. We repeat a
    // similar process for the last visible column.
```

**Review by farmaazon — APPROVED** [review]
Please add a pivotal issue for making the ideal solution for column width handling. I can fill the details about data structure later. Then, add info that resizing columns may have performance problems with large number of columns (with link to the issue).

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 19]
```diff
@@ -510,6 +510,11 @@ pub struct GridViewTemplate<
 /// After either method of resizing, each visible entry in the affected column will receive
 /// the [`EntryFrp::set_size`] event. It is up to the entry implementation to avoid loops between
 /// [`EntryFrp::set_size`] and [`EntryFrp::override_column_width`].
+///
+/// The current implementation has performance implications for large amounts of entries. A more
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]:
```suggestion
/// **Important**: The current implementation has performance implications for large amounts of entries. A more
```

**Review by MichaelMauderer — APPROVED** [review]
